core/src/main/java/io/grpc/LoadBalancer2.java

```diff
@@ -0,0 +1,327 @@
+  }
+
+  /**
+   * A balancing decision made by {@link SubchannelPicker} for an RPC.
+   */
+  @Immutable
+  public static final class PickResult {
+    // A READY channel, or null
+    @Nullable private final Subchannel subchannel;
+    // An error to be propagated to the application if subchannel == null
+    // Or OK if there is no error.
+    // subchannel being null and error being OK means RPC needs to wait
+    private final Status status;
+
+    private PickResult(Subchannel subchannel, Status status) {
+      this.subchannel = subchannel;
+      this.status = Preconditions.checkNotNull(status, "status");
+    }
+
+    /**
+     * A decision to proceed the RPC on a Subchannel.  The state of the Subchannel is supposed to be
+     * {@link ConnectivityState#READY}.  However, a non-READY Subchannel will not fail the RPC,
+     * but will only leave it buffered.
+     */
+    public static PickResult withSubchannel(Subchannel subchannel) {
+      return new PickResult(Preconditions.checkNotNull(subchannel, "subchannel"), Status.OK);
+    }
+
+    /**
+     * A decision to report a connectivity error to the RPC.  If the RPC is {@link
+     * CallOptions#withWaitForReady wait-for-ready}, it will stay buffered.  Otherwise, it will fail
+     * with the given error.
+     *
+     * @param error the error status.  Must not be OK.
+     */
+    public static PickResult withError(Status error) {
+      Preconditions.checkArgument(!error.isOk(), "error status shouldn't be OK");
+      return new PickResult(null, error);
+    }
+
+    /**
+     * No decision could be made.  The RPC will stay buffered.
+     */
+    public static PickResult withNothingYet() {
+      return new PickResult(null, Status.OK);
+    }
+
+    /**
+     * The Subchannel if this result was created by {@link #withSubchannel}, or null otherwise.
+     */
+    @Nullable
+    public Subchannel getSubchannel() {
+      return subchannel;
+    }
+
+    /**
+     * The status associated with this result.  Non-{@code OK} if created with {@link #withError},
+     * or {@code OK} otherwise.
+     */
+    public Status getStatus() {
+      return status;
+    }
+
+    @Override
+    public String toString() {
+      return "[subchannel=" + subchannel + " status=" + status + "]";
+    }
+  }
+
+  /**
+   * Provides essentials for LoadBalancer implementations.
+   */
+  @ThreadSafe
+  public abstract static class Helper {
+    /**
+     * Creates a Subchannel, which is a logic connection to the given group of addresses which are
+     * considered equivalent.  The {@code attrs} are custom attributes associated with this
+     * Subchannel, and can be accessed later through {@link Subchannel#getAttributes}.
+     *
+     * <p>The LoadBalancer is responsible for closing unused Subchannels, and closing all
+     * Subchannels within {@link #shutdown}.
+     */
+    public abstract Subchannel createSubChannel(EquivalentAddressGroup addrs, Attributes attrs);
```

Comment:
nit: `createSubChannel` => `createSubchannel` to make it consistent with other class names